server/chat-plugins/youtube.ts

```diff
@@ -41,19 +43,36 @@ export class YoutubeInterface {
 		const res = JSON.parse(raw);
 		if (!res || !res.items || res.items.length < 1) return;
 		const data = res.items[0];
-		const cache = {
-			name: data.snippet.title,
-			description: data.snippet.description,
-			url: data.snippet.customUrl,
-			icon: data.snippet.thumbnails.medium.url,
-			videos: Number(data.statistics.videoCount),
-			subs: Number(data.statistics.subscriberCount),
-			views: Number(data.statistics.viewCount),
-			username: username,
-		};
-		this.data[id] = {...cache};
+		if (!this.data[id]) {
+			const cache = {
+				name: data.snippet.title,
+				description: data.snippet.description,
+				url: data.snippet.customUrl,
+				icon: data.snippet.thumbnails.medium.url,
+				videos: Number(data.statistics.videoCount),
+				subs: Number(data.statistics.subscriberCount),
+				views: Number(data.statistics.viewCount),
+				username: username,
+			};
+			this.data[id] = {...cache};
+		} else {
+			const cached = this.data[id];
+			// convert to arrays, if old
+			if (!Array.isArray(cached.subs)) cached.subs = [cached.subs];
+			if (!Array.isArray(cached.views)) cached.views = [cached.views];
+			// add new data to the beginning
+			cached.subs.unshift(Number(data.statistics.subscriberCount));
+			cached.views.unshift(Number(data.statistics.viewCount));
+			cached.subs.slice(0, 15);
+			cached.views.slice(0, 15);
+			// update all the rest
+			for (const key in data) {
+				if (Array.isArray(cached[key])) continue;
+				cached[key] = data[key];
+			}
```

Comment:
These should be done one at a time, for safety.

Reply:
i.e. Don't use `for...in` on an AnyObject or otherwise bypass TypeScript's safety checks.